lib/amazonka/src/Amazonka/Auth/SSO.hs

```diff
@@ -68,33 +67,21 @@ fromSSO cachedTokenFile ssoRegion accountId roleName env = do
       -- The Region you SSO through may differ from the Region you intend to
       -- interact with after. The former is handled here, the latter is taken
       -- care of later, in ConfigFile.
-      let ssoEnv = env {envRegion = ssoRegion}
-
-      resp <-
-        runResourceT $
-          sendUnsigned ssoEnv $
-            SSO.newGetRoleCredentials roleName accountId (fromSensitive accessToken)
+      let ssoEnv = env {envRegion = ssoRegion, envAuth = Proxy}
```

Comment:
So I see here we are accepting any `Env' withAuth` and so we `envAuth = Proxy` to be able to use `sendUnsigned` at `Env' Proxy`. Could that clobbering be risky in that someone might mis-use `fromSSO` with an authorized `Env` unintentionally? Or is that a feature?

I'm wondering if moving the type-restriction up as necessary, e.g.

```hs
fromSSO ::
  (MonadIO m, Foldable withAuth) =>
  FilePath ->
  Region ->
  -- | Account ID
  Text ->
  -- | Role Name
  Text ->
  Env' Proxy ->
  m Env
```

Would be better.

Reply:
I thought about it after reading your comment, but I don't think so. When we provide `Env' withAuth` in a type signature, we're saying to the caller "I don't care whether or not this env has auth information", which I think is a pretty clear signal that it's not going to be used. But `Amazonka.HTTP.httpRequest` uses the `Foldable` instance to get auth information if it is present, so I have to put `Env' Proxy` into the type signature of `sendUnsigned`, otherwise it would sign requests if auth information was present - not really what we want to do.

So that forces `sendUnsigned` to take an `Env' Proxy` argument, but why not push that explicit `Proxy` into the type signature of functions like `fromSSO`? My main fear is that it makes it needlessly harder to build credential chains: consider some code which might try to `sts:AssumeRoleWithWebIdentity` (on ECS) or `sts:AssumeRole` (when a dev is working locally). The first method would need an `Env' Proxy`, and the second would need an `Env`, and they'd no longer unify. I think it's better to just ignore the auth information on any `Env'` we are given.

It does remind me of a few tidy-ups that should happen as part of this PR: we have an `EnvNoAuth` type alias which should be used, and `fromSSO` and `fromWebIdentity` should be able to drop the `Foldable withAuth` constraints.